### PR TITLE
new: usr: HADTWO-1973: Namenode out of safe mode

### PR DESCRIPTION
**Pull request by kartik-bhatia**
Provided a function that attempts to bring out namenode out of safe mode.

### CODE REVIEW

File: hadoop/util.sh [hunk 1]
```diff
@@ -73,3 +73,36 @@ function use_java8() {
...
+# second : Number of seconds each attempt will sleep for waiting for namenode to come out of sleep mode
+function wait_until_namenode_out_of_safe() {
+    n=0
+    num_attempts=$1
+    sleep_sec=$2
+    nn_out_of_safe_mode=0
+    until [ $n -ge $attempts ]
+    do
+        n=$[$n+1]
+        safe_mode_stat=`sudo -u hdfs hadoop dfsadmin -safemode get|awk '{print $4}'`
```
Comment (hiyer) [inline hadoop/util.sh]: sudo should not be required. bootstraps run as root.

File: hadoop/util.sh [hunk 2]
```diff
@@ -73,3 +73,36 @@ function use_java8() {
    restart_worker_services
  fi
 }
+
+# Wait until namenode is out of safe mode. Takes 2 params
+# first : Number of attempts function will make to get namenode out of safemode
+# second : Number of seconds each attempt will sleep for waiting for namenode to come out of sleep mode
+function wait_until_namenode_out_of_safe() {
+    n=0
+    num_attempts=$1
```
Comment (hiyer) [inline hadoop/util.sh]: We should have a default value for these fields - say 5s for sleep and 50 for number of retries?

**Review by hiyer — CHANGES_REQUESTED** [review]

File: hadoop/util.sh [hunk 3]
```diff
@@ -73,3 +73,45 @@ function use_java8() {
...
+
+# Wait until namenode is out of safe mode. Takes 2 optional params
+# first : Number of attempts function will make to get namenode out of safemode. Default is 50
+# second : Number of seconds each attempt will sleep for waiting for namenode to come out of sleep mode. Default is 5 sec
+function wait_until_namenode_out_of_safe() {
+    n=0
+    num_attempts=$1
+    sleep_sec=$2
+    
+    if [ -z "$num_attempts" ]; then
```
Comment (hiyer) [inline hadoop/util.sh]: Nit - can change initialization to num_attempts=${1:-num_attempts} and similarly for sleep_sec.

File: hadoop/util.sh [hunk 4]
```diff
@@ -73,3 +73,45 @@ function use_java8() {
...
+            echo "Attempt $n/$attempts"
+            sleep $sleep_sec
+        else
+            echo "NN is out of safemode..."
+            nn_out_of_safe_mode=1
+            break
+        fi
+    done
+    if [[ $nn_out_of_safe_mode -eq 0 ]]; then
+        safe_mode_stat=`sudo -u hdfs hadoop dfsadmin -safemode get|awk '{print $4}'`
```
Comment (hiyer) [inline hadoop/util.sh]: Why do we have this additional check after the loop?

**Review by hiyer — CHANGES_REQUESTED** [review]

File: hadoop/util.sh (hunk 4 above)
Comment (kartik-bhatia) [reply hadoop/util.sh]: To Check if the last attempt made was successful to bring namenode out of safemode or not

**Review by hiyer — APPROVED** [review]

**Pull request merged**